File: sdk/kotlin/src/test/kotlin/com/loomup/client/Mocks.kt
package com.loomup.client

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import java.nio.charset.StandardCharsets
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

// MARK: - HTTP mock

class MockHttp : HttpTransport {
    data class Call(
        val method: String,
        val url: String,
        val auth: String?,
        val body: ByteArray?,
    )

    private val calls = CopyOnWriteArrayList<Call>()
    var handler: (suspend (method: String, url: String, auth: String?, body: ByteArray?) -> Pair<ByteArray, Int>)? =
        null

    override suspend fun execute(request: HttpRequest): HttpResponse {
        val auth = request.headers["Authorization"]
        calls.add(Call(request.method, request.url, auth, request.body))
        val h = handler ?: throw LoomupError("no mock handler", code = "test")
        val (data, status) = h(request.method, request.url, auth, request.body)
        return HttpResponse(status = status, body = data)
    }

    fun snapshotCalls(): List<Call> = calls.toList()
}

fun jsonBytes(json: String): ByteArray = json.toByteArray(StandardCharsets.UTF_8)

// MARK: - WebSocket mock

class MockWebSocket : WebSocketConnecting {
    override var onOpen: (() -> Unit)? = null
    override var onMessage: ((String) -> Unit)? = null
    override var onClose: (() -> Unit)? = null

    private val sent = CopyOnWriteArrayList<String>()
    private val _isOpen = AtomicBoolean(false)
    private val _connecting = AtomicBoolean(false)
    private val connectCount = AtomicInteger(0)

    var openDelayMs: Long = 0
    /** If false, stay CONNECTING forever (for timeout tests). */
    var autoOpen: Boolean = true
    /** Optional hook after each outbound frame. */
    var onSend: ((String) -> Unit)? = null

    override val isOpen: Boolean
        get() = _isOpen.get()

    override val isConnectingOrOpen: Boolean
        get() = _connecting.get() || _isOpen.get()

    val connectCountValue: Int
        get() = connectCount.get()

    override fun connect(url: String) {
        connectCount.incrementAndGet()
        _connecting.set(true)
        _isOpen.set(false)
        if (!autoOpen) return
        if (openDelayMs <= 0) {
            simulateOpen()
        } else {
            Thread {
                Thread.sleep(openDelayMs)
                simulateOpen()
            }.start()
        }
    }

    override fun send(text: String) {
        sent.add(text)
        onSend?.invoke(text)
    }

    override fun close() {
        _isOpen.set(false)
        _connecting.set(false)
        onClose?.invoke()
    }

    fun simulateOpen() {
        _isOpen.set(true)
        _connecting.set(false)
        onOpen?.invoke()
    }

    fun simulateMessage(text: String) {
        onMessage?.invoke(text)
    }

    fun simulateClose() {
        _isOpen.set(false)
        _connecting.set(false)
        onClose?.invoke()
    }

    fun snapshotSent(): List<String> = sent.toList()

    fun clearSent() {
        sent.clear()
    }

    fun parsedSent(): List<JsonObject> {
        val json = Json { ignoreUnknownKeys = true }
        return snapshotSent().mapNotNull { text ->
            try {
                json.parseToJsonElement(text).jsonObject
            } catch (_: Exception) {
                null
            }
        }
    }
}

/** Holds the latest mock socket so tests can drive messages after connect. */
class MockWebSocketBox {
    private val current = AtomicReference<MockWebSocket?>(null)
    var autoOpen: Boolean = true
    var openDelayMs: Long = 0

    fun note(ws: MockWebSocket) {
        current.set(ws)
    }

    fun factory(): WebSocketFactory = {
        val ws = MockWebSocket()
        ws.autoOpen = autoOpen
        ws.openDelayMs = openDelayMs
        note(ws)
        ws
    }

    val socket: MockWebSocket?
        get() = current.get()
}

File: swift/Tests/LoomupTests/Mocks.swift
import Foundation
@testable import Loomup

// MARK: - HTTP mock

final class MockHTTP: HTTPTransport, @unchecked Sendable {
    struct Call: Sendable {
        let method: String
        let url: String
        let auth: String?
        let body: Data?
    }

    private let lock = NSLock()
    private(set) var calls: [Call] = []
    /// Handler receives method, url string, auth header, body.
    var handler: (@Sendable (String, String, String?, Data?) async throws -> (Data, Int))?

    func data(for request: URLRequest) async throws -> (Data, URLResponse) {
        let method = request.httpMethod ?? "GET"
        let url = request.url?.absoluteString ?? ""
        let auth = request.value(forHTTPHeaderField: "Authorization")
        let body = request.httpBody
        lock.lock()
        calls.append(Call(method: method, url: url, auth: auth, body: body))
        lock.unlock()
        guard let handler else {
            throw LoomupError("no mock handler", code: "test")
        }
        let (data, status) = try await handler(method, url, auth, body)
        let response = HTTPURLResponse(
            url: request.url ?? URL(string: "http://example.test")!,
            statusCode: status,
            httpVersion: nil,
            headerFields: ["Content-Type": "application/json"]
        )!
        return (data, response)
    }

    func snapshotCalls() -> [Call] {
        lock.lock()
        defer { lock.unlock() }
        return calls
    }
}

func jsonData(_ object: Any) -> Data {
    try! JSONSerialization.data(withJSONObject: object)
}

// MARK: - WebSocket mock

final class MockWebSocket: WebSocketConnecting, @unchecked Sendable {
    var onOpen: (() -> Void)?
    var onMessage: ((String) -> Void)?
    var onClose: (() -> Void)?

    private let lock = NSLock()
    private(set) var sent: [String] = []
    private(set) var connectCount = 0
    private var _isOpen = false
    private var _connecting = false
    var openDelay: TimeInterval = 0
    /// If false, stay CONNECTING forever (for timeout tests).
    var autoOpen = true
    /// Optional hook after each outbound frame (used to auto-ack/error in tests).
    var onSend: ((String) -> Void)?

    var isOpen: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _isOpen
    }

    var isConnectingOrOpen: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _connecting || _isOpen
    }

    func connect(url: URL) {
        lock.lock()
        connectCount += 1
        _connecting = true
        _isOpen = false
        lock.unlock()
        guard autoOpen else { return }
        let delay = openDelay
        DispatchQueue.global().asyncAfter(deadline: .now() + delay) { [weak self] in
            self?.simulateOpen()
        }
    }

    func send(text: String) {
        lock.lock()
        sent.append(text)
        let hook = onSend
        lock.unlock()
        hook?(text)
    }

    func close() {
        lock.lock()
        _isOpen = false
        _connecting = false
        lock.unlock()
        onClose?()
    }

    func simulateOpen() {
        lock.lock()
        _isOpen = true
        _connecting = false
        lock.unlock()
        onOpen?()
    }

    func simulateMessage(_ text: String) {
        onMessage?(text)
    }

    func simulateClose() {
        lock.lock()
        _isOpen = false
        _connecting = false
        lock.unlock()
        onClose?()
    }

    func snapshotSent() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        return sent
    }

    func clearSent() {
        lock.lock()
        sent.removeAll()
        lock.unlock()
    }

    func parsedSent() -> [[String: Any]] {
        snapshotSent().compactMap { text in
            guard let data = text.data(using: .utf8),
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { return nil }
            return obj
        }
    }
}

/// Holds the latest mock socket so tests can drive messages after connect.
final class MockWebSocketBox: @unchecked Sendable {
    private let lock = NSLock()
    private var current: MockWebSocket?
    var autoOpen = true
    var openDelay: TimeInterval = 0

    func note(_ ws: MockWebSocket) {
        lock.lock()
        current = ws
        lock.unlock()
    }

    func factory() -> WebSocketFactory {
        { [weak self] in
            let ws = MockWebSocket()
            ws.autoOpen = self?.autoOpen ?? true
            ws.openDelay = self?.openDelay ?? 0
            self?.note(ws)
            return ws
        }
    }

    var socket: MockWebSocket? {
        lock.lock()
        defer { lock.unlock() }
        return current
    }
}
